Title: Unpivoting Columns 
Slug: unpivoting_columns 
Summary: Reverse pivot columns to be added as new rows in a dataframe. 
Date: 2016-11-27 12:00 
Category: wrangling_data 
Tags: 
Authors: Nate Hall

<a href="http://nbviewer.jupyter.org/github/nathan-hall/nathan-hall.github.io/blob/pelican/content/wrangling_data/editing_column_names.ipynb" target="_blank">Link to the full jupyter notebook.</a><br/>
<a href="http://nbviewer.jupyter.org/github/nathan-hall/nathan-hall.github.io/blob/pelican/content/wrangling_data/editing_column_names_code.ipynb" target="_blank">Link to the code only jupyter notebook.</a>

- We'll be using the billboard hot 100 dataset as the example.
- We wanted to get the dataset to have rows representing the lowest level of data possible.
- <a href="./wrangling_data/editing_column_names.html", target="_blank">We cleaned up the column names in this article.</a>
- Now to finish the the final step to accomplish the goal below.
## Goal: To get each row to represent the week performance of a song.

In [11]:
import pandas as pd

In [12]:
data = pd.read_csv('../data/billboard_weeks_edited.csv')
df = pd.DataFrame(data)
df.head(1)

year           artist                     track  time genre     entered  \
0  2000  Destiny's Child  Independent Women Part I  3:38  Rock  2000-09-23   

       peaked   1     2     3 ...  67  68  69  70  71  72  73  74  75  76  
0  2000-11-18  78  63.0  49.0 ... NaN NaN NaN NaN NaN NaN NaN NaN NaN NaN  

[1 rows x 83 columns]

In [14]:
#Check the shape of the dataset we're starting with.
df.shape

(317, 83)

We know that we have 317 rows and 83 columns. If we are "unpivoting" this to be fewer columns we can expect the rows to increase substantially. Remember this for when we check the shape again later.

## Melt its face off!
We will use the ".melt" function to accomplish what we need for this dataset. More information on it can be found in the <a href="http://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html", target="_blank"> documentation here.</a>

### Notes on pd.melt
First off we see that it is not applied to a dataframe, so we don't do "df.melt" it actually gets passed the entire dataframe object as an argument and is called by starting with "pd."
```python
pd.melt(df)
```

## Create a new dataframe
Simply for the ease of still being able to track back to our original csv in case something doesn't work. We will then declare a new dataframe that will take on the results of the pd.melt( ) function.
```python
df2 = pd.melt(df)
```

## id_vars=[ ]
We already passed the dataframe. The id_vars argument will be the columns that will be retained in the dataframe so we'll declare that by listing the columns we want to keep. Note that this argument is taking on a list so we can pass the columns to it with all the columns list methods that we know. For this example i'll just list out the columns since there are so few.
```python
df2 = pd.melt(df, id_vars=['year', 'artist', 'track', 'time', 'genre', 'entered', 'peaked'])
```
## value_vars=[ ]
This argument says it is for stating all the columns to "unpivot". Phew, we have a lot of weeks... do we have to list them all? That's the beauty of this function. If we simply leave this blank it will default to unpivot all of the columns that are not listed in the "id_vars=[ ]" argument. SWEET!!!
## var_name=[ ]
So we have this tricky thing to do with unpivoting... we'll have all the week columns get turned into row cells and they will need a column header. To do this we declare the "var_name=[ ]" argument with the column header we want.
```python
df2 = pd.melt(df, id_vars=['year', 'artist', 'track', 'time', 'genre', 'entered', 'peaked'], var_name='week')
```
## value_name[ ]
Now what about all the values that are listed in each week column right now? Where are those supposed to go? Glad you asked that. These are the values that we are "unpivoting" but they to will need a column header since we just took the week columns away. We know that those values represent the weekly ranking of a song during that week so we'll "unpivot" those values to a new column called "rank"

In [13]:
df2 = pd.melt(df, id_vars=['year', 'artist', 'track', 'time', 'genre', 'entered', 'peaked'],
              var_name='week', value_name='rank')

Check your work.

In [15]:
df2.head()

year               artist                                  track  time  \
0  2000      Destiny's Child               Independent Women Part I  3:38   
1  2000              Santana                           Maria, Maria  4:18   
2  2000        Savage Garden                     I Knew I Loved You  4:07   
3  2000              Madonna                                  Music  3:45   
4  2000  Aguilera, Christina  Come On Over Baby (All I Want Is You)  3:38   

  genre     entered      peaked week  rank  
0  Rock  2000-09-23  2000-11-18    1  78.0  
1  Rock  2000-02-12  2000-04-08    1  15.0  
2  Rock  1999-10-23  2000-01-29    1  71.0  
3  Rock  2000-08-12  2000-09-16    1  41.0  
4  Rock  2000-08-05  2000-10-14    1  57.0

## Cool! It worked!
We now have the rows representing the lowest level of data in our dataset. What did this do to the shape of our table? Remember what it was at the beginning? Lets check it again to see what happened.

In [16]:
df2.shape

(24092, 9)

## Woah. That is MUCH longer.
Nothing to worry about though. You'll find that pandas can handle this length of a data table easily. Most importantly we have our dataset setup in a way where we can work with it much better.